File: bindings/rust/src/safe_bloom.rs
// SPDX-License-Identifier: PMPL-1.0-or-later
// Copyright (c) 2026 Jonathan D.A. Jewell (hyperpolymath) <jonathan.jewell@open.ac.uk>

//! Safe Bloom filter via libproven FFI.
//!
//! Provides probabilistic set membership testing.
//! All operations delegate to Idris 2 verified code.

use crate::core::{Error, Result};
use crate::ffi;

/// A Bloom filter for probabilistic set membership.
///
/// A Bloom filter can tell you definitively that an element is NOT in the
/// set, or that it MIGHT be in the set (with a tunable false positive rate).
///
/// The underlying memory is managed by libproven. The filter is freed
/// when this struct is dropped.
pub struct BloomFilter {
    ptr: *mut ffi::BloomFilter,
}

// SAFETY: The underlying bloom filter is managed by libproven.
unsafe impl Send for BloomFilter {}

impl BloomFilter {
    /// Create a new Bloom filter.
    ///
    /// - `expected_elements`: expected number of elements to insert.
    /// - `false_positive_rate`: desired false positive rate (e.g., 0.01 for 1%).
    pub fn new(expected_elements: usize, false_positive_rate: f64) -> Result<Self> {
        // SAFETY: proven_bloom_create takes value-type arguments;
        // always safe to call.
        let ptr = unsafe {
            ffi::proven_bloom_create(expected_elements, false_positive_rate)
        };
        if ptr.is_null() {
            return Err(Error::AllocationFailed);
        }
        Ok(BloomFilter { ptr })
    }

    /// Add an element to the filter.
    pub fn add(&mut self, data: &[u8]) {
        // SAFETY: self.ptr is valid (checked at construction), and we pass
        // a valid pointer and length from a Rust slice.
        unsafe {
            ffi::proven_bloom_add(self.ptr, data.as_ptr(), data.len());
        }
    }

    /// Add a string element to the filter.
    pub fn add_str(&mut self, s: &str) {
        self.add(s.as_bytes());
    }

    /// Check if an element might be in the filter.
    ///
    /// Returns `true` if the element might be present (could be a false
    /// positive), `false` if it is definitely not present.
    pub fn contains(&self, data: &[u8]) -> bool {
        // SAFETY: self.ptr is valid (checked at construction), and we pass
        // a valid pointer and length from a Rust slice.
        unsafe {
            ffi::proven_bloom_contains(self.ptr, data.as_ptr(), data.len())
        }
    }

    /// Check if a string element might be in the filter.
    pub fn contains_str(&self, s: &str) -> bool {
        self.contains(s.as_bytes())
    }
}

impl Drop for BloomFilter {
    fn drop(&mut self) {
        // SAFETY: self.ptr was allocated by proven_bloom_create
        // and has not been freed yet.
        unsafe {
            ffi::proven_bloom_free(self.ptr);
        }
    }
}

File: bindings/zig/src/safe_bloom.zig
// SPDX-License-Identifier: PMPL-1.0-or-later
// Copyright (c) 2026 Jonathan D.A. Jewell (hyperpolymath) <jonathan.jewell@open.ac.uk>
//
// Proven SafeBloom - FFI bindings to libproven Bloom filter operations.
// All computation is performed in verified Idris 2 code via libproven.

const std = @import("std");
const c = @cImport(@cInclude("proven.h"));

/// Error types for Bloom filter operations.
pub const BloomError = error{
    CreationFailed,
};

/// Bloom filter backed by libproven.
pub const BloomFilter = struct {
    ptr: *c.ProvenBloomFilter,

    /// Add element to Bloom filter via libproven.
    pub fn add(self: BloomFilter, data: []const u8) void {
        c.proven_bloom_add(self.ptr, data.ptr, data.len);
    }

    /// Check if element might be in filter via libproven.
    /// Returns true if probably present (may be false positive),
    /// false if definitely absent.
    pub fn contains(self: BloomFilter, data: []const u8) bool {
        return c.proven_bloom_contains(self.ptr, data.ptr, data.len);
    }

    /// Free Bloom filter via libproven.
    pub fn deinit(self: BloomFilter) void {
        c.proven_bloom_free(self.ptr);
    }
};

/// Create a Bloom filter via libproven.
/// expected_elements: Expected number of elements.
/// false_positive_rate: Desired false positive rate (0 < rate < 1).
pub fn create(expected_elements: usize, false_positive_rate: f64) BloomError!BloomFilter {
    const ptr = c.proven_bloom_create(expected_elements, false_positive_rate);
    if (ptr == null) return error.CreationFailed;
    return BloomFilter{ .ptr = ptr.? };
}

test "create and add/contains" {
    const bf = try create(1000, 0.01);
    defer bf.deinit();
    bf.add("hello");
    bf.add("world");
    try std.testing.expect(bf.contains("hello"));
    try std.testing.expect(bf.contains("world"));
    try std.testing.expect(!bf.contains("definitely not present xyz123"));
}
